tests/functional_tests/rpc_payment.py

```diff
@@ -0,0 +1,411 @@
+        entry = data[0]
+        assert entry.rpc == 'rpc_access_tracking'
+        assert entry.count == 1
+
+    def test_access_mining(self):
+        print('Testing access mining')
+        daemon = Daemon(idx=1)
+        wallet = Wallet(idx=3)
+
+        res = daemon.rpc_access_info(client = self.get_signature())
+        assert len(res.hashing_blob) > 39
+        assert res.height == 1
+        assert res.top_hash == '418015bb9ae982a1975da7d79277c2705727a56894ba0fb246adaabb1f4632e3'
+        assert res.credits_per_hash_found == 5000
+        assert res.diff == 10
+        assert res.credits == 0
+        cookie = res.cookie
+
+        # Try random nonces till we find one that's valid and one that's invalid
+        nonce = 0
+        found_valid = 0
+        found_invalid = 0
+        last_credits = 0
+        while found_valid == 0 or found_invalid == 0:
+            nonce += 1
+            try:
+                res = daemon.rpc_access_submit_nonce(nonce = nonce, cookie = cookie, client = self.get_signature())
+                found_valid += 1
+                assert res.credits == last_credits + 5000
+            except Exception as e:
+                found_invalid += 1
+                res = daemon.rpc_access_info(client = self.get_signature())
+                assert res.credits < last_credits or res.credits == 0
+            assert nonce < 1000 # can't find both valid and invalid -> the RPC probably fails
+            last_credits = res.credits
+
+        # we should now have 1 valid nonce, and a number of bad ones
+        res = daemon.rpc_access_info(client = self.get_signature())
+        assert len(res.hashing_blob) > 39
+        assert res.height > 1
+        assert res.top_hash != '418015bb9ae982a1975da7d79277c2705727a56894ba0fb246adaabb1f4632e3' # here, any share matches network diff
+        assert res.credits_per_hash_found == 5000
+        assert res.diff == 10
+        cookie = res.cookie
+
+        res = daemon.rpc_access_data()
+        assert len(res.entries) > 0
+        e = [x for x in res.entries if x['client'] == self.public_key]
+        assert len(e) == 1
+        e = e[0]
+        assert e.nonces_stale == 0
+        assert e.nonces_bad == found_invalid
+        assert e.nonces_good == found_valid
+        assert e.nonces_dupe == 0
+        assert e.nonces_dupe == 0
+
+        # Try random nonces till we find one that's valid so we get a load of credits
+        while last_credits == 0:
+            nonce += 1
+            try:
+                res = daemon.rpc_access_submit_nonce(nonce = nonce, cookie = cookie, client = self.get_signature())
+                found_valid += 1
+                break
+            except:
+                found_invalid += 1
+            assert nonce < 1000 # can't find both valid and invalid -> the RPC probably fails
+
+        # we should now have 5000
+        res = daemon.rpc_access_info(client = self.get_signature())
+        assert res.credits == 5000
```

Comment:
I can see this holds when `last_credits` is zero, but what about the other case? Can't `res.credits` be somewhere above 0 and not equal to 5000?

Reply:
The loop's actually wrong, it doesn't update last_credits. I'll go through it again.

Reply:
It has to be >= 5000 since last nonce was valid.